src/audio.rs

```diff
@@ -0,0 +1,86 @@
+/// and return the mixed audio
+pub fn process_twilio_media(
+    media: twilio_response::EventMedia,
+    mut buffer_data: &mut BufferData,
+) -> Option<Vec<u8>> {
+    // NOTE: when Twilio sends media data, it should send 20 ms audio chunks at a time, where each ms of audio is 8 bytes
+    let media_chunk = base64::decode(media.payload).unwrap();
+    let media_chunk_size = media_chunk.len();
+    if media_chunk_size != 20 * 8 {
+        // WARN: Twilio media chunk size is not the expected size of 20 * 8 bytes.
+    }
+    // NOTE: I've seen cases where the timestamp is less than 20 ms ahead of the previous chunk
+    let timestamp = media.timestamp.parse::<u32>().unwrap();
+
+    if media.track == "inbound" {
+        let time_lost = if timestamp < buffer_data.inbound_last_timestamp + 20 {
+            // WARN: Received inbound timestamp is less than 20 ms ahead of previous timestamp.
+            0
+        } else {
+            (timestamp - (buffer_data.inbound_last_timestamp + 20))
+                .try_into()
+                .unwrap()
+        };
+        if time_lost > 0 {
+            // silence for mulaw data is 0xff, and there are 8 bytes of mulaw data per ms
+            let silence = &vec![0xff; 8 * time_lost];
+            buffer_data.inbound_buffer.extend_from_slice(&silence[..]);
+        }
+        buffer_data
+            .inbound_buffer
+            .extend_from_slice(&media_chunk[..]);
+        buffer_data.inbound_last_timestamp = timestamp;
+    } else if media.track == "outbound" {
+        let time_lost = if timestamp < buffer_data.outbound_last_timestamp + 20 {
+            // WARN: Received outbound timestamp is less than 20 ms ahead of previous timestamp.
+            0
+        } else {
+            (timestamp - (buffer_data.outbound_last_timestamp + 20))
+                .try_into()
+                .unwrap()
+        };
+        if time_lost > 0 {
+            // silence for mulaw data is 0xff, and there are 8 bytes of mulaw data per ms
+            let silence = &vec![0xff; 8 * time_lost];
+            buffer_data.outbound_buffer.extend_from_slice(&silence[..]);
+        }
+        buffer_data
+            .outbound_buffer
+            .extend_from_slice(&media_chunk[..]);
+        buffer_data.outbound_last_timestamp = timestamp;
+    }
+
+    // we will send audio to deepgram once we have 400 ms (20 * 20 * 8 bytes) of audio (or more)
+    let mut mixed = Vec::new();
+    while buffer_data.inbound_buffer.len() >= 20 * 20 * 8
+        && buffer_data.outbound_buffer.len() >= 20 * 20 * 8
+    {
+        let inbound_buffer_segment: Vec<u8> =
+            buffer_data.inbound_buffer.drain(0..20 * 20 * 8).collect();
+        let outbound_buffer_segment: Vec<u8> =
+            buffer_data.outbound_buffer.drain(0..20 * 20 * 8).collect();
+
+        for sample in 0..20 * 20 * 8 {
+            mixed.push(inbound_buffer_segment[sample]);
+            mixed.push(outbound_buffer_segment[sample]);
+        }
```

Comment:
Could save on some allocations and indexing and such with some iterator stuff. Also ends up a bit shorter :shrug: 
```suggestion
        let inbound_buffer_segment = buffer_data.inbound_buffer.drain(0..20 * 20 * 8);
        let outbound_buffer_segment = buffer_data.outbound_buffer.drain(0..20 * 20 * 8);

        for (inbound, outbound) in inbound_buffer_segment.zip(outbound_buffer_segment) {
            mixed.push(inbound);
            mixed.push(outbound);
        }
```

Reply:
woah